process/setup_sp.py

```diff
@@ -373,6 +373,100 @@ def process_distant_nodes(
     return(distant_nodes)
 
 
+#Cumulative opportunities (binary)
+#1 if d <= access_dist
+#0 if d > access_dist
+def binary_access_score(df, new_name, old_name, access_dist):
+    """
+    Calculate accessibiity score using binary measure: 1 if access <= access_dist, 0 otherwise
+
+    Parameters
+    ----------
+    df: DataFrame
+        DataFrame with origin-destination distances
+    new_name: list
+        list of new field names for new access score
+    old_name: list
+         list of original field names for distances
+    access_dist: int
+        access distance threshold
+
+    Returns
+    -------
+    DataFrame
+    """
+    df[new_name] = (df[old_name] <= access_dist).astype("Int64").fillna(0)
+    return df[new_name]
+
+
+#Soft threshold access score
+#Higgs, C., Badland, H., Simons, K. et al. The Urban Liveability Index
+def soft_access_score(df, new_name, old_name, access_dist, k):
+    """
+    Calculate accessibiity score using soft threshold approach:
+    1 / (1+ e ^(k *((dist-access_dist)/access_dist)))
+
+    Parameters
+    ----------
+    df: DataFrame
+        DataFrame with origin-destination distances
+    new_name: list
+        list of new field names for new access score columns
+    old_name: list
+         list of original field names for distance columns
+    access_dist: int
+        access distance threshold
+    k: int
+        the slope of decay
+
+    Returns
+    -------
+    DataFrame
+    """
+    df[new_name] = 1 / (1+numpy.exp(k * ((df[old_name]-access_dist) / access_dist)))
+
+    df[new_name] = df[new_name].astype(float).fillna(0)
+    return df[new_name]
+
+#Cumulative-Gaussian
+#Reference: Vale, D. S., & Pereira, M. (2017).
+#The influence of the impedance function on gravity-based pedestrian accessibility measures
+def Cumulative_Gaussian_access_score(df, new_name, old_name, access_dist, k):

```

Comment:
I think we should provide some default values to guide the user as to possible sensible parameterisation of k; e.g.
def Cumulative_Gaussian_access_score(df, new_name, old_name, access_dist=500, k=5)
I'm not sure if k=5 is sensible for this formula for a distance of 500 metres, but if it might be handy having some kind of suggestion of an appropriate decay for 500 metres (ie. if we were to use it for our case study, what would be appropriate?)

I'm not sure why it is necessary to loop through columns in this function, but I think there is a problem in the logic --- in my tests, it seems to replace all values of the 'old name' variable in the input dataframe with 0 as a result of running the function. I am not certain how best to parameterise the decay, but so far (and possibly because all inputs are replaced as zero), the derived scores all evaluate as 1.0 for me? I had a look to see if I can refactor, but perhaps I'm confused; any thoughts?

Here is my test of the basic case where a distance of 700m was recorded relative to a target threshold of 500; intuitively, this should score something below 1 and probably above 0.5.  I struggled to get it to approach that without using extreme values for k; I found 100,000 worked as a usable value in the end though:

```
for k in [.0001,.001,.01,.1,1,10,100,1000,10000,100000,1000000]:
...   print(f"k={k:10}:\t{numpy.exp(-1 * (((700-500)**2) / k))}")
```
...
k=    0.0001:   0.0
k=     0.001:   0.0
k=      0.01:   0.0
k=       0.1:   0.0
k=         1:   0.0
k=        10:   0.0
k=       100:   1.9151695967140057e-174
k=      1000:   4.248354255291589e-18
k=     10000:   0.01831563888873418
k=    100000:   0.6703200460356393
k=   1000000:   0.9607894391523232

so, using k=100,000, perhaps this function could be an alternative, avoiding the looping (which seemed to be implemented in a way that was destructive of the input source in my tests)

```
def Cumulative_Gaussian_access_score(df, distances, access_dist=500, k=100000):
    """
    Calculate accessibiity score using Cumulative-Gaussian approach:
    1 if d <= access_dist ; otherwise, e ^(-1 *((d^2)/k)) if d > access_dist
    Parameters
    ----------
    df: DataFrame
        DataFrame with origin-destination distances
    distances: list
        list of field names for distance records
    access_dist: int
        access distance threshold
    k: int
        the slope of decay
    Returns
    -------
    DataFrame
    """
    df1 = df[distances].copy()
    df1[df1<=access_dist] = 1
    df1[df1>access_dist] = numpy.exp(-1 * (((df1[df1>access_dist]-access_dist)**2) / k))
    df1 = df1.astype(float).fillna(0)
    return(df1)
```

An example of implementation is

```
k=100000
df[[f'cum_{x}' for x in distances]] = Cumulative_Gaussian_access_score(df, distances, access_dist, k)
df
```

This yields the following, adding onto the dataframe with soft threshold inputs defined earlier,
```

     x1    x2   soft_x1   soft_x2    cum_x1    cum_x2
0   100  1000  0.982014  0.006693  1.000000  0.082085
1   200   900  0.952574  0.017986  1.000000  0.201897
2   300   800  0.880797  0.047426  1.000000  0.406570
3   400   700  0.731059  0.119203  1.000000  0.670320
4   500   600  0.500000  0.268941  1.000000  0.904837
5   600   500  0.268941  0.500000  0.904837  1.000000
6   700   400  0.119203  0.731059  0.670320  1.000000
7   800   300  0.047426  0.880797  0.406570  1.000000
8   900   200  0.017986  0.952574  0.201897  1.000000
9  1000   100  0.006693  0.982014  0.082085  1.000000

```

Reply:
> I'm not sure why it is necessary to loop through columns in this function, but I think there is a problem in the logic --- in my tests, it seems to replace all values of the 'old name' variable in the input dataframe with 0 as a result of running the function. 

I used for loop to loop through each column, as I remember in my previous test, I was not able to produce a boolean mask for the entire dataframe. Within the for loop, It first creates new columns of 'new name' variables with value 0, then proceed with calculation, I suppose it should not replace the 'old name' variable with 0"


> Here is my test of the basic case where a distance of 700m was recorded relative to a target threshold of 500; intuitively, this should score something below 1 and probably above 0.5. I struggled to get it to approach that without using extreme values for k; I found 100,000 worked as a usable value in the end though:

Thanks for checking on the parameters, I think it is a good idea to keep default parameters, for this Cumulative_Gaussian function, I would refer to Vale and Pereira (2016) https://journals.sagepub.com/doi/10.1177/0265813516641685; they suggested use k=129,842 for accessibility of 500m, perhaps we could this number as default parameters?



I just tested your functions using the sample point dataframe in Vic, but it returns me with the error: 

```
c_m_names = [f"{x.replace('nearest_node','access')}_c_m" for x in distance_names]
samplePointsData[c_m_names] = Cumulative_Gaussian_access_score(samplePointsData, distance_names, parameters['accessibility_distance'])
samplePointsData[c_m_names]
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-135-f909b77ceb1c> in <module>()
----> 1 samplePointsData[[f'cum_{x}' for x in distances]] = Cumulative_Gaussian_access_score(samplePointsData, distance_names, access_dist, k)

<ipython-input-127-182d031e8ddb> in Cumulative_Gaussian_access_score(df, distances, access_dist, k)
     18     """
     19     df1 = df[distances].copy()
---> 20     df1[df1<=access_dist] = 1
     21     df1[df1>access_dist] = numpy.exp(-1 * (((df1[df1>access_dist]-access_dist)**2) / k))
     22     df1 = df1.astype(float).fillna(0)

/opt/conda/lib/python3.7/site-packages/pandas/core/frame.py in __setitem__(self, key, value)
   3037 
   3038         if isinstance(key, DataFrame) or getattr(key, "ndim", None) == 2:
-> 3039             self._setitem_frame(key, value)
   3040         elif isinstance(key, (Series, np.ndarray, list, Index)):
   3041             self._setitem_array(key, value)

/opt/conda/lib/python3.7/site-packages/pandas/core/frame.py in _setitem_frame(self, key, value)
   3086         if key.size and not is_bool_dtype(key.values):
   3087             raise TypeError(
-> 3088                 "Must pass DataFrame or 2-d ndarray with boolean values only"
   3089             )
   3090 

TypeError: Must pass DataFrame or 2-d ndarray with boolean values only
```

Reply:
Hmm.... that's odd; I can't check with the actual data right now (some IT issues with RMIT, so can't log in to work computer), but using what should be equivalent test data in the code below it appears to work,
```
>>> def cumulative_gaussian_access_score(df, distances, threshold=500, k=129842):
...     """
...     Calculate accessibiity score using Cumulative-Gaussian approach:
...     1 if d <= threshold ; otherwise, e ^(-1 *((d^2)/k)) if d > threshold
...     Parameters
...     ----------
...     df: DataFrame
...         DataFrame with origin-destination distances
...     distances: list
...         list of field names for distance records
...     threshold: int
...         access distance threshold
...     k: int
...         the slope of decay
...     Returns
...     -------
...     DataFrame
...     """
...     distances = df[distances].copy()
...     distances[distances<=threshold] = 1
...     distances[distances>threshold] = numpy.exp(-1 * (((distances[distances>threshold]-threshold)**2) / k))
...     scores = distances.fillna(0)
...     return scores
...
>>>
>>>
>>> df = pd.DataFrame({'x1':[100,numpy.NaN,300,400,500,600,700,800,900,1000],'x2':[1000,900,800,700,600,500,400,300,200,100]})
>>>
>>> distances = ['x1','x2']
>>> df[[f'access_{x}' for x in distances]] = cumulative_gaussian_access_score(df,distances,threshold)
>>> df
       x1    x2  access_x1  access_x2
0   100.0  1000   1.000000   0.145815
1     NaN   900   0.000000   0.291631
2   300.0   800   1.000000   0.499999
3   400.0   700   1.000000   0.734866
4   500.0   600   1.000000   0.925874
5   600.0   500   0.925874   1.000000
6   700.0   400   0.734866   1.000000
7   800.0   300   0.499999   1.000000
8   900.0   200   0.291631   1.000000
9  1000.0   100   0.145815   1.000000
```

Reply:
The test data also work for me too, but still not working for the actual data, I am checking to see what would be the issue there...

Reply:
That's really strange... I'm not really sure why it works in the example but fails in the data... but you're right, it has to work in the data!  If you step through line by line is it more clear what isn't working with the Vic data? e.g. for the example which works (with Pandas 1.2.2, Python 3.8),
```
import pandas as pd
import numpy

df = pd.DataFrame({'x1':[100,numpy.NaN,300,400,500,600,700,800,900,1000],'x2':[1000,900,800,700,600,500,400,300,200,100]})
distances = ['x1','x2'] 
threshold=500

distances = df[distances].copy()
distances[distances<=threshold] = 1
distances[distances>threshold] = numpy.exp(-1 * (((distances[distances>threshold]-threshold)**2) / k))
scores = distances.fillna(0)
```
I'll try to check it out with the actual data and code later in the week.

Reply:
My best guess is the problem with the missing values, it seems to work when I do:
```
df1 = samplePointsData[distance_names].copy()
mask = (df1<=access_dist).fillna(False) 
mask = mask.astype(bool) # have to change this to bool, otherwise it return me "object" data type and not work as boolean mask
df1[mask]
```

I will check back tomorrow on this, and will keep you updated if all work :)

Reply:
I think I found a fix to the problem, this is related to the data types. Now the function reads:
```
#Cumulative-Gaussian
#Reference: Vale, D. S., & Pereira, M. (2017).
#The influence of the impedance function on gravity-based pedestrian accessibility measures

def cumulative_gaussian_access_score(df, distance_names, threshold=500, k=129842):
    """
    Calculate accessibiity score using Cumulative-Gaussian approach:
    1 if d <= access_dist ; otherwise, e ^(-1 *((d^2)/k)) if d > access_dist

    Parameters
    ----------
    df: DataFrame
        DataFrame with origin-destination distances
    distance_names: list
        list of field names for distance records
    threshold: int
        access distance threshold
    k: int
        the slope of decay

    Returns
    -------
    DataFrame
    """
    df1 = df[distance_names].copy()
    df1 = df1.astype(float)
    df1[df1<=threshold] = 1
    df1[df1>threshold] = numpy.exp(-1 * (((df1[df1>threshold]-threshold)**2) / k))
    df1 = df1.fillna(0).astype(float)
    return df1
```
By adding `df1 = df1.astype(float)` seem to resolve the issue, this works when testing with Vic data